### PR TITLE
[AMORO-2111] Support list partitions by filter via flink catalog

### LINKED ISSUE #2111: [Improvement]: Support list partitions by filter via flink catalog
### Search before asking

- [X] I have searched in the [issues](https://github.com/NetEase/amoro/issues?q=is%3Aissue) and found no similar issues.

### What would you like to be improved?

Now the flink catalog is not implemented `public List<CatalogPartitionSpec> listPartitionsByFilter(ObjectPath tablePath,List<Expression> filters)` interface; so SHOW PARTITIONS tbl PARTITION ('xxx'='xxx') statement is not supported.

### How should we improve?

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Subtasks

_No response_

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by huyuanfeng2018**
## Why are the changes needed?

Close #2111.

## Brief change log

- impl `listPartitionsByFilter(ObjectPath tablePath, List<Expression> filters)` and `listPartitions(ObjectPath tablePath, CatalogPartitionSpec partitionSpec)`

## How was this patch tested?

- [X] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [X] Run test locally before making a pull request

## Documentation

- Does this pull request introduce a new feature? (no)
- If yes, how is the feature documented? (not documented)

### CODE REVIEW

File: flink/flink-common/src/main/java/com/netease/arctic/flink/catalog/ArcticCatalog.java [hunk 1]
```diff
@@ -578,4 +606,29 @@ public InternalCatalogBuilder catalogBuilder() {
...
+   * @param arcticPartitionSpec arcticPartitionSpec
+   * @param tablePath tablePath
+   * @throws PartitionSpecInvalidException thrown if any key in partitionSpec doesn't exist in
+   *     partitionKeys.
+   */
+  private void checkValidPartitionSpec(
+      CatalogPartitionSpec partitionSpec, PartitionSpec arcticPartitionSpec, ObjectPath tablePath)
+      throws PartitionSpecInvalidException {
+    for (PartitionField field : arcticPartitionSpec.fields()) {
+      if (!partitionSpec.getPartitionSpec().containsKey(field.name())) {
```
Comment (YesOrNo828) [inline flink/flink-common/src/main/java/com/netease/arctic/flink/catalog/ArcticCatalog.java:623]: I think we should check that `arcticPartitionSpec.fields()` contains all the `partitionSpec.getPartitionSpec().keySet()`
```suggestion
      List<String> partitionKeys = spec.fields().stream().map(PartitionField::name).collect(Collectors.toList());
      for (String key : partitionSpec.getPartitionSpec().keySet()) {
            if (!partitionKeys.contains(key)) {
                throw new PartitionSpecInvalidException(
                        getName(), partitionKeys, tablePath, partitionSpec);
            }
        }
```

**Review by baiyangtx — APPROVED** [review]
LGTM,  @YesOrNo828  PTAL.

File: flink/flink-common/src/main/java/com/netease/arctic/flink/catalog/ArcticCatalog.java (hunk 1 above)
Comment (huyuanfeng2018) [reply flink/flink-common/src/main/java/com/netease/arctic/flink/catalog/ArcticCatalog.java]: fixd.

**Review by YesOrNo828 — APPROVED** [review]
LGTM. After the CI passes, I'll merge into the master.

**Pull request merged**